pydantic_ai_slim/pydantic_ai/toolsets/_searchable.py

```diff
@@ -0,0 +1,185 @@
+        # If any nested SearchableToolset instances have added their own search tools, drop those so that only one
+        # search tool is exposed to the model.
+        tools = {name: tool for name, tool in tools.items() if not is_search_tool(tool.tool_def)}
+
+        all_tools: dict[str, ToolsetTool[AgentDepsT]] = {}
+
+        for tool_name, tool in tools.items():
+            defer = tool.tool_def.defer_loading
+            all_tools[tool_name] = _SearchToolsetToolWrapper[AgentDepsT](tool, self) if defer else tool
+
+        search_tool, search_toolset_tool = _search_tool(toolset=self)
+
+        if search_tool.name in all_tools:
+            raise UserError(
+                f'Tool name {search_tool.name!r} is reserved by Pydantic AI for implementing defer_loading=True tools. '
+                f'Rename the tool or wrap the toolset in a `PrefixedToolset` to avoid the conflict.'
+            )
+        all_tools[search_tool.name] = search_toolset_tool
+
+        return all_tools
+
+    async def call_tool(
+        self, name: str, tool_args: dict[str, Any], ctx: RunContext[AgentDepsT], tool: ToolsetTool[AgentDepsT]
+    ) -> Any:
+        if is_search_tool(tool.tool_def):
+            return await self.search_tools(ctx, tool_args['regex'])
+        elif isinstance(tool, _SearchToolsetToolWrapper):
+            return await self.wrapped.call_tool(name, tool_args, ctx, tool.wrapped)
+        else:
+            return await self.wrapped.call_tool(name, tool_args, ctx, tool)
+
+    async def search_tools(self, ctx: RunContext[AgentDepsT], regex: str) -> list[str]:
+        """Searches for tools matching the query, activates them and returns their names."""
+        toolset_tools = await self.wrapped.get_tools(ctx)
+        matching_tool_names: list[str] = []
+
+        rx = re.compile(regex)
+
+        for _, tool in toolset_tools.items():
+            desc = tool.tool_def.description
+            if rx.search(tool.tool_def.name) or desc is not None and rx.search(desc):
+                matching_tool_names.append(tool.tool_def.name)
+
+        self._active_tool_names.update(matching_tool_names)
+        return matching_tool_names
+
+    def is_active(self, tool_def: ToolDefinition) -> bool:
+        return tool_def.name in self._active_tool_names
+
+
+def _search_tool(toolset: SearchableToolset) -> tuple[Tool, ToolsetTool[AgentDepsT]]:
+    async def search(ctx: RunContext[AgentDepsT], regex: str) -> list[str]:
+        return await toolset.search_tools(ctx, regex)
+
+    # TODO Check if pattern is a better name than regex.
+    # TODO Are examples allowed to be defined somewhere to expose to the model?
+    schema = {
+        'type': 'object',
+        'properties': {
+            'regex': {
+                'type': 'string',
+                'description': 'Regex pattern to search for relevant tools',
+            }
+        },
+        'required': ['regex'],
+    }
+
+    desc = 'Search and load additional tools that defer tool loading'
```

Comment:
Need to be able to change based on how intelligent the model is.  Per model defaults?